File: USACO.GUIDE/SILVER/Binary Search/USACO_594.cpp
#include <bits/stdc++.h>

using namespace std;

typedef int64_t ll;

int check(vector<ll> arr, ll R, ll K){

    for(ll i = 0; i < K; i++){
        ll start, end, m = 0;
        for(ll j = 0; j < arr.size(); j++){
            ll left, right = 0;
            for(ll t0 = j; t0 < arr.size(); t0++ ){
                if (arr[t0] <= arr[j]+R && arr[t0] >= arr[j]-R){
                    right = t0;
                }
            }
            for(ll t1 = j; t1 >= 0; t1--){
                if (arr[t1] <= arr[j]+R && arr[t1] >= arr[j]-R){
                    left = t1;
                }

            }
            if (right-left+1 > m){
                start = left;
                end = right;
            } 
        }
    }
    
}

int main(){
    

    ll n, k;
    vector<ll> arr;

    scanf("%lld %lld", &n, &k);

    for(ll i = 0; i < n; i++){
        arr.push_back(0);
        scanf("%lld", &arr.back());
    }

    sort(arr.begin(), arr.end());

    //binary search

    
    return 0;
}

File: USACO.GUIDE/SILVER/Binary Search/USACO_594.py
from bisect import bisect_right as br

input = open('angry.in', 'r').readline
out = open('angry.out', 'w')

def check(arr, R, K):
    start = 0
    for i in range(K):
        idx = br(arr, arr[start]+R*2) - 1
        start = idx + 1
        if start >= len(arr):
            return 1

    if start == len(arr):
        return 1
    else:
        return 0

n, k = map(int, input().split())
arr =[]

for _ in range(n):
    arr.append(int(input()))

arr.sort()

#Binary Search
hi = 2*10**9
lo = 0
while(lo < hi):
    mid = (lo+hi)//2
    if check(arr, mid, k):
        hi = mid
    else:
        lo = mid + 1

print(hi, file=out)